app/components/edit/EditServiceChildCollection.tsx

```diff
@@ -0,0 +1,120 @@
+import React, {
+  FC, useState, useRef, useEffect,
+} from 'react';
+
+/**
+ * Renders a child property collection of a Service as a list of individual components,
+ * called ResourceObjectItem. Enables the user to add, modify, and delete members of
+ * the collection and save to the DB.
+ *
+ * In the future, this component could possibly be abstracted to replace the
+ * EditCollection HOC class-based JSX component that does essentially the same thing
+ * for child property collections of Resources.
+ */
+interface ComponentProps {
+  index: number;
+  item: any;
+  handleItemChange: any;
+}
+
+export const EditServiceChildCollection = ({
+  initialCollectionData,
+  handleCollectionChange,
+  ResourceObjectItem,
+  label,
+  blankTemplateObj,
+  buttonText,
+  propertyKeyName,
+}: {
+  initialCollectionData: any[];
+  handleCollectionChange: (field: any, value: any) => null;
+  ResourceObjectItem: FC<componentProps>;
+  label: string;
+  blankTemplateObj: object;
+  buttonText: string;
+  propertyKeyName: string;
+}) => {
+  const [resourceCollection, setResourceCollection] = useState(initialCollectionData || []);
+  const firstUpdate = useRef(true);
+  const updateParentService = useRef(true);
+
+  useEffect(() => {
+    if (firstUpdate.current) {
+      firstUpdate.current = false;
+      return;
+    }
+
+    if (updateParentService.current) {
+      // We only want to update the parent service when an item is deleted or modified;
+      // we don't want to call update when an empty template has been added
+      handleCollectionChange(propertyKeyName, resourceCollection);
+    }
```

Comment:
It's a little bit strange for me to see something like this in a `useEffect()` call. It's a bit hard for me to put it clearly in words, but I think this is working against the fundamental model of React of state and state updates. Most React components encapsulate state via the `useState()` hook (or this.state for class-based components), and they encapsulate state updates with function calls (`setState()` for many kinds of updates, and arbitrary function callbacks for more complex state updates that you may want to encapsulate).

What you have here, instead, is almost an inversion of the model, where you are listening in to specific changes to the state rather than to the individual state update calls. I think this is why you ended up having to define some ex-React persistent state via refs.

Have you considered a different approach where you directly trigger calls to `handleCollectionChange()` in the individual handler methods you've defined here? For example, if you don't want to update the parent collection when adding a new item, you can simply omit the `handleCollectionChange()` call in that case, but in the other two cases explicitly call it. I think this would make the component a bit easier to reason about, since instead of having to reason about it at two layers — the immediate handler functions and then the effect handler —, everything collapses into the immediate handler functions.

Reply:
I can't quite remember why I did it this way. I think I did initially try to call the `handleCollectionChange` function within the individual handler functions. But I may have been passing the `resourceCollection` state var into the `handleCollectionChange` function rather than the updated `newCollection`. The `resourceCollection` state var wasn't being updated before the `handleCollectionChange` executing, thus I created this whole roundabout way of doing this. 

Anyhow, invoking the `handleCollectionChange` within the individual handlers and passing in the `newCollection` var works as desired. It also removes quite a bit of code, in addition to following the principles you've identified above. Thanks for calling this out!